python/ray/train/jax/jax_trainer.py

```diff
@@ -0,0 +1,160 @@
+from typing import Callable, Optional, Dict, Union, TYPE_CHECKING
+import logging
+
+from ray.air._internal.config import ensure_only_allowed_dataclass_keys_updated
+from ray.train.jax.config import JaxConfig
+from ray.train.trainer import GenDataset
+from ray.train.data_parallel_trainer import DataParallelTrainer
+from ray.air.config import ScalingConfig, RunConfig, DatasetConfig
+from ray.air.checkpoint import Checkpoint
+from ray.util import PublicAPI
+
+if TYPE_CHECKING:
+    from ray.data.preprocessor import Preprocessor
+
+logger = logging.getLogger(__name__)
+
+
+@PublicAPI(stability="alpha")
+class JaxTrainer(DataParallelTrainer):
+    """A Trainer for data parallel jax training.
+
+    This Trainer runs the function ``train_loop_per_worker`` on multiple Ray
+    Actors. These actors already have the necessary jax process group already
+    configured for distributed jax training.
+
+    The ``train_loop_per_worker`` function is expected to take in either 0 or 1
+    arguments:
+
+    .. code-block:: python
+
+        def train_loop_per_worker():
+            ...
+
+    .. code-block:: python
+
+        def train_loop_per_worker(config: Dict):
+            ...
+
+    If ``train_loop_per_worker`` accepts an argument, then
+    ``train_loop_config`` will be passed in as the argument. This is useful if you
+    want to tune the values in ``train_loop_config`` as hyperparameters.
+
+    If the ``datasets`` dict contains a training dataset (denoted by
+    the "train" key), then it will be split into multiple dataset
+    shards that can then be accessed by ``ray.train.get_dataset_shard("train")`` inside
+    ``train_loop_per_worker``. All the other datasets will not be split and
+    ``ray.train.get_dataset_shard(...)`` will return the the entire Dataset.
+
+    Inside the ``train_loop_per_worker`` function, you can use any of the
+    :ref:`Ray Train function utils <train-api-func-utils>`.
```

Comment:
Should we go into a bit more detail here about what sort of a cluster/scaling config Jax needs to work and explain what magic we are doing in that regard if any? and also talk about the CPU/placement strategy stuff. Basically anything we warn or throw an exception about should be in the docstring. We don't want to make the users do trial and error

Also, do we still require users to import jax inside their training unction? we should mention that too if it's still the case

Reply:
got it, makes a lot of sense to me, thanks! will add these!

Reply:
sounds good